class-voicewp-setup.php

```diff
@@ -0,0 +1,200 @@
+	 *
+	 * @param int $post_id Post ID
+	 * @param object $post Post object
+	 * @param bool $update Is post updated or new
+	 */
+	public function save_post( $post_id, $post, $update ) {
+		if (
+			empty( $post_id )
+			|| wp_is_post_revision( $post_id )
+			|| ( defined( 'WP_IMPORTING' ) && WP_IMPORTING === true )
+			|| ! in_array( $post->post_type, voicewp_news_post_types() )
+			|| 'publish' !== get_post_status( $post_id )
+			|| false == $update
+		) {
+			return;
+		}
+		delete_transient( 'voicewp_single_' . $post_id );
+	}
+
+	/**
+	 * Handle clearing cache on single items when a post is edited
+	 *
+	 * @param int $post_id Post ID
+	 * @param object $post Post object
+	 * @param bool $update Is post updated or new
+	 */
+	public function publish_clear_cache( $post_id, $post ) {
+		delete_transient( 'voicewp_latest' );
+		$args = array(
+			'post_type' => voicewp_news_post_types(),
+			'posts_per_page' => 5,
+			'tax_query' => array(),
+		);
+		$news = new \Alexa\Skill\News;
+		$news->endpoint_content( $args );
+	}
+
+	/**
+	 * Add image sizes for Alexa cards.
+	 */
+	public function add_image_size() {
+		add_image_size( 'alexa-small', 720, 480, true );
+		add_image_size( 'alexa-large', 1200, 800, true );
+	}
+
+	/**
+	 * Add the Alexa service to the list of allowed HTTP origins
+	 *
+	 * @param $allowed_origins array Default allowed HTTP origins
+	 * @return array allowed origin URLs
+	 */
+	public function allowed_http_origins( $allowed_origins ) {
+		$allowed_origins[] = 'http://ask-ifr-download.s3.amazonaws.com';
+		$allowed_origins[] = 'https://ask-ifr-download.s3.amazonaws.com';
+		return $allowed_origins;
+	}
+
+	/**
+	 * Action hook callback for plugins_loaded.
+	 * Sets a plugin version.
+	 * Migration functions will never run on new installs.
+	 *
+	 * @since 1.0.0
+	 */
+	public function plugins_loaded() {
+		// Determine if the database version and code version are the same.
+		$current_version = get_option( 'voicewp_version' );
+		if ( version_compare( $current_version, self::$version, '>=' ) ) {
+			return;
+		}
+		/**
+		 * Determine if this is a legacy version
+		 * If this option is empty, it's a new install, or not legacy
+		 * If the option exists, trigger the update functions
+		 */
+		$settings = get_option( 'alexawp-settings' );
+		if ( ! empty( $settings ) ) {
+			$this->upgrade_to_1_0_0();
+		}
+
+		// Set the database version to the current version in code.
+		update_option( 'voicewp_version', self::$version );
+	}
+
+	/**
+	 * Upgrades options, meta and post types to be version 1.0.0 compatible
+	 *
+	 * @access public
+	 */
+	public function upgrade_to_1_0_0() {
+		$this->upgrade_post_types_and_meta();
+		$this->upgrade_options();
+	}
+
+	/**
+	 * Attempt to migrate meta and post types from an older version of this plugin.
+	 * Changes the prefix of options, meta, post types
+	 *
+	 * @access public
+	 */
+	public function upgrade_post_types_and_meta() {
```

Comment:
This could theoretically time out if there were a lot of posts and it ran longer than 30 seconds. Probably not likely given the usage of this plugin thus far.

Looking at this, I honestly think you should scrap the whole upgrade path and just release this as a separate plugin. This isn't without precedent - FB IA did this between v1 and v2 and that's in far greater use.

Any upgrade function that loops over posts is dangerous so I'd consider this option strongly. Anyone on AlexaWP would need to manually upgrade to VoiceWP. Maybe VoiceWP could have a CLI to handle this, but I would not automate this functionality on upgrade.